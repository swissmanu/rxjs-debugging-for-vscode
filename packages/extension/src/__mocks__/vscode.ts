/**
 * This file contains a set of fragile mocks of vscode API objects. They allow us to run Jest-based unit tests for our
 * own code.
 * These are only used for unit tests. Integration tests are executed using the official extension test runner.
 */

import type { Event as EventType, EventEmitter as EventEmitterType, Uri as UriType } from 'vscode';

export class Position {
  constructor(readonly line: number, readonly character: number) {}
}

export class EventEmitter<T> implements EventEmitterType<T> {
  private listeners: ((data: T) => void)[] = [];

  event: EventType<T> = (listener) => {
    this.listeners.push(listener);
    return new Disposable(() => {});
  };

  /**
   * Notify all subscribers of the {@link EventEmitter.event event}. Failure
   * of one or more listener will not fail this function call.
   *
   * @param data The event object.
   */
  fire(data: T): void {
    for (const listener of this.listeners) {
      listener(data);
    }
  }

  /**
   * Dispose this object and free resources.
   */
  dispose(): void {}
}

export class Disposable {
  /**
   * Combine many disposable-likes into one. Use this method
   * when having objects with a dispose function which are not
   * instances of Disposable.
   *
   * @param disposableLikes Objects that have at least a `dispose`-function member.
   * @return Returns a new disposable which, upon dispose, will
   * dispose all provided disposables.
   */
  static from(...disposableLikes: { dispose: () => any }[]): Disposable {
    return new Disposable(() => {});
  }

  /**
   * Creates a new Disposable calling the provided function
   * on dispose.
   * @param callOnDispose Function that disposes something.
   */
  constructor(callOnDispose: () => void) {}

  /**
   * Dispose this object.
   */
  dispose(): void {}
}

export class Uri implements UriType {
  static parse(value: string, strict?: boolean): Uri {
    return new Uri('', '', '', '', '');
  }

  static file(path: string): Uri {
    return new Uri('', '', '', '', '');
  }

  static joinPath(base: Uri, ...pathSegments: string[]): Uri {
    return new Uri('', '', '', '', '');
  }

  static from(components: {
    scheme: string;
    authority?: string;
    path?: string;
    query?: string;
    fragment?: string;
  }): Uri {
    return new Uri('', '', '', '', '');
  }

  private constructor(
    readonly scheme: string,
    readonly authority: string,
    readonly path: string,
    readonly query: string,
    readonly fragment: string
  ) {}

  readonly fsPath: string = '';

  with(change: { scheme?: string; authority?: string; path?: string; query?: string; fragment?: string }): Uri {
    return this;
  }

  toString(skipEncoding?: boolean): string {
    return '';
  }

  toJSON(): any {
    return {};
  }
}
