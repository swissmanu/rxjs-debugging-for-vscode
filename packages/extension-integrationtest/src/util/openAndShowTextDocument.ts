import * as vscode from 'vscode';

export default async function openAndShowTextDocument(filePattern: string): Promise<vscode.TextDocument> {
  const [file] = await vscode.workspace.findFiles(filePattern);
  const document = await vscode.workspace.openTextDocument(file);
  await vscode.window.showTextDocument(document);

  return document;
}
