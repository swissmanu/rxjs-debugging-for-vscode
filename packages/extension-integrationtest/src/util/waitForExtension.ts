import { Commands } from 'rxjs-debugging-for-vs-code/out/integrationTest';
import * as vscode from 'vscode';

async function waitForCommands(): Promise<void> {
  const command = Commands.EnableOperatorLogPoint;
  const commands = await vscode.commands.getCommands(true);
  if (commands.indexOf(command) !== -1) {
    return;
  }
  throw new Error(`"${command}" Command not found`);
}

async function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export default async function waitForExtension(): Promise<void> {
  let ok = false;
  for (let retry = 0; retry < 5; retry++) {
    try {
      await waitForCommands();
      ok = true;
    } catch (_) {
      await wait(1000);
    }
  }

  if (!ok) {
    throw new Error('Could not determine if extension was available or not.');
  }
}
