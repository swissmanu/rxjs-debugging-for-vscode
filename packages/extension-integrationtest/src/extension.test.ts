import {
  Commands,
  executeCommand,
  OperatorLogPoint,
  TestCommands,
} from 'rxjs-debugging-for-vs-code/out/integrationTest';
import * as vscode from 'vscode';
import openAndShowTextDocument from './util/openAndShowTextDocument';
import waitForExtension from './util/waitForExtension';

describe('RxJS Debugging for vscode', () => {
  test('shows operator life cycle events as text editor decoration', async () => {
    const document = await openAndShowTextDocument('**/commonjs/observable.js');
    await waitForExtension();

    // Enable Operator Log Point for the first operator, take.
    await executeCommand(
      vscode.commands,
      Commands.EnableOperatorLogPoint,
      new OperatorLogPoint(
        document.uri,
        new vscode.Position(5, 4),
        {
          fileName: document.uri.fsPath,
          line: 5,
          character: 25,
          operatorIndex: 0,
        },
        'take'
      )
    );

    const debuggingDone = new Promise<void>((resolve) => {
      vscode.debug.onDidTerminateDebugSession(() => resolve());
    });
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    await vscode.debug.startDebugging(vscode.workspace.workspaceFolders![0], 'Launch CommonJS');
    await debuggingDone;

    const decorations = await executeCommand(
      vscode.commands,
      TestCommands.GetDecorationSetterRecording,
      'src/commonjs/observable.js',
      'liveLog'
    );

    expect(decorations).toMatchSnapshot();
  });
});
