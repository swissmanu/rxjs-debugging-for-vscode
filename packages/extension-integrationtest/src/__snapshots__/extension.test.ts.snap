// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RxJS Debugging for vscode shows operator life cycle events as text editor decoration 1`] = `
Array [
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Next: 0\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Next: 1\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Next: 2\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Next: 3\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Completed\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Unsubscribe\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Unsubscribe\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [
      "(5:9007199254740991):(5:9007199254740991)-undefined-{\\"after\\":{\\"contentText\\":\\"Unsubscribe\\"}}",
    ],
    "ranges": Array [],
  },
  Object {
    "decorationType": "TextEditorDecorationType8",
    "options": Array [],
    "ranges": Array [],
  },
]
`;
